src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ImmutableArrayExtensions.cs

```diff
@@ -41,19 +41,47 @@ internal static ImmutableArray<T> ToImmutableArrayOrEmpty<T>(this IEnumerable<T>
             return ImmutableArray.CreateRange<T>(items);
         }
 
-        internal static ImmutableArray<T> ToImmutableArrayOrEmpty<T>(this ImmutableArray<T> items)
-            => items.IsDefault ? ImmutableArray<T>.Empty : items;
-
-        internal static IReadOnlyList<T> ToImmutableReadOnlyListOrEmpty<T>(this IEnumerable<T>? items)
+        internal static IReadOnlyList<T> ToBoxedImmutableArray<T>(this IEnumerable<T>? items)
         {
-            if (items is ImmutableArray<T> array && !array.IsDefault)
+            if (items is null)
+            {
+                return SpecializedCollections.EmptyBoxedImmutableArray<T>();
+            }
+
+            if (items is ImmutableArray<T> array)
+            {
+                return array.IsDefaultOrEmpty ? SpecializedCollections.EmptyBoxedImmutableArray<T>() : (IReadOnlyList<T>)items;
+            }
+
+            if (items is ICollection<T> collection && collection.Count == 0)
             {
-                return (IReadOnlyList<T>)items;
+                return SpecializedCollections.EmptyBoxedImmutableArray<T>();
             }
-            else
+
+            return ImmutableArray.CreateRange(items);
+        }
+
+        /// <summary>
+        /// Use to validate public API input for properties that are exposed as <see cref="IReadOnlyList{T}"/>.
+        /// 
+        /// Pattern:
+        /// <code>
+        /// argument.AsBoxedImmutableArrayWithNonNullItems() ?? throw new ArgumentNullException(nameof(argument)),
```

Comment:
so given that all callers must do the `throw ...`, it makes more sense for me for this to push teh `throw` into this method.  Can we do that?  There's no requirement that the 'throw' happen in the first frame the user is calling into.  There are tons of argument-checking helpers in the bcl that push the arg-check+throw into the helper.

Reply:
I don't like pushing it down, except for helpers that are dedicated argument validation helpers for mutliple reasons:

1) It's not clear whether a public API is actually validating all of its arguments if it leaves it up to the methods it calls to do so.
2) When writing an internal method, how do I decide whether or not to do full argument validation? It's inconsistent and confusing if some internal methods of a type do argument validation because they happen to be called from some public API impl, but others don't because they currently aren't called from public API impl. 
3) The methods called from the public API implementation might have different argument names than the public API. They might also have more or less arguments than the public API. Throwing an argument exception with wrong argument name is not very user friendly - it defeats the purpose of having argument name on the exception.
4) Internal calls to the internal methods would do redundant argument validation.

Reply:
I'm not too concerned about repeating code that does argument validation in multiple public API entry-points. That's the cost of exposing public API. It's also not a problem to write a dedicated argument validation helpers that help reduce the duplication. But then they should just be simple validation helpers.

Reply:
Yes.  I'm saying: make this one of those "simple validation helpers" :)  That's what it feels liek to me.  It takes in the arg (and the name of hte arg), validates it and passed the result through in a sanitized form that hte rest of hte system can handle :)

It's not a big deal here.  however, it seems sensible since the literal 100% expectation (and this code demosntrates it) is that it is used only in scenarios with arg-checking.

Reply:
Ah, I see what you mean now. Yes, that would be ok. I guess it's worth the tiny overhead of looking up the parameter name string on every call in this case.

Reply:
@CyrusNajmabadi and I tried several alternative patterns and concluded none is ideal, so I'm gonna keep it as is.